Mathlib/FieldTheory/SeparableDegree.lean

```diff
@@ -268,31 +250,543 @@ def embProdEmbOfIsAlgebraic [Algebra E K] [IsScalarTower F E K] (halg : Algebra.
+  rw [← finSepDegree_top, ← finrank_top F E]
+  refine induction_on_adjoin P ?_ (fun L x h ↦ ?_) ⊤
+  · simp only [finSepDegree_bot, IntermediateField.finrank_bot]
+  simp only at h ⊢
+  have heq : _ * _ = _ * _ := congr_arg₂ (· * ·) h <|
+    (finSepDegree_adjoin_simple_eq_finrank_iff L E x (IsAlgebraic.of_finite L x)).2 <|
+      (IsSeparable.separable F x).map (f := algebraMap F L) |>.of_dvd
+        (minpoly.dvd_map_of_isScalarTower F L x)
+  set M := L⟮x⟯; clear_value M
+  letI : Algebra L M := Subalgebra.algebra M.toSubalgebra
+  letI : Module L M := Algebra.toModule
+  letI : SMul L M := Algebra.toSMul
+  haveI : IsScalarTower F L M := IntermediateField.isScalarTower M
+  rwa [finSepDegree_mul_finSepDegree_of_isAlgebraic F L M (Algebra.IsAlgebraic.of_finite L M),
+    FiniteDimensional.finrank_mul_finrank F L M] at heq
+
+/-- If `E / F` is a finite extension, then its separable degree is equal to its degree if and
+only if it is a separable extension. -/
+theorem finSepDegree_eq_finrank_iff [FiniteDimensional F E] :
+    finSepDegree F E = finrank F E ↔ IsSeparable F E :=
+  ⟨fun heq ↦ ⟨fun x ↦ by
+    have halg := IsAlgebraic.of_finite F x
+    refine (finSepDegree_adjoin_simple_eq_finrank_iff F E x halg).1 <| le_antisymm
+      (finSepDegree_adjoin_simple_le_finrank F E x halg) <| le_of_not_lt fun h ↦ ?_
+    have := Nat.mul_lt_mul h (finSepDegree_le_finrank F⟮x⟯ E) Fin.size_pos'
+    rw [finSepDegree_mul_finSepDegree_of_isAlgebraic F F⟮x⟯ E (Algebra.IsAlgebraic.of_finite _ E),
+      FiniteDimensional.finrank_mul_finrank F F⟮x⟯ E] at this
+    linarith only [heq, this]⟩, fun _ ↦ finSepDegree_eq_finrank_of_isSeparable F E⟩
+
+end Field
+
+namespace IntermediateField
+
+lemma separable_of_mem_isSeparable {L : IntermediateField F E} [IsSeparable F L]
+    {x : E} (h : x ∈ L) : IsIntegral F x ∧ (minpoly F x).Separable :=
+  ⟨isIntegral_iff.1 <| IsSeparable.isIntegral F (K := L) ⟨x, h⟩, by
+    simpa only [minpoly_eq] using IsSeparable.separable F (K := L) ⟨x, h⟩⟩
+
+/-- `F⟮x⟯ / F` is a separable extension if and only if `x` is a separable element.
+As a consequence, any rational function of `x` is also a separable element. -/
+theorem isSeparable_adjoin_simple_iff_separable {x : E} :
+    IsSeparable F F⟮x⟯ ↔ IsIntegral F x ∧ (minpoly F x).Separable := by
+  refine ⟨fun hsep ↦ ?_, fun ⟨h, hsep⟩ ↦ ?_⟩
+  · exact separable_of_mem_isSeparable F E <| mem_adjoin_simple_self F x
+  · haveI := adjoin.finiteDimensional h
+    rwa [← finSepDegree_eq_finrank_iff,
+      finSepDegree_adjoin_simple_eq_finrank_iff F E x h.isAlgebraic]
+
+/-- If `x` and `y` are both separable elements, then `F⟮x, y⟯ / F` is a separable extension.
+As a consequence, any rational function of `x` and `y` is also a separable element. -/
+theorem isSeparable_adjoin_two_of_separable {x y : E}
+    (hx : IsIntegral F x ∧ (minpoly F x).Separable)
+    (hy : IsIntegral F y ∧ (minpoly F y).Separable) :
+    IsSeparable F F⟮x, y⟯ := by
```

Comment:
It's probably unnecessary to carry the IsIntegral conditions around after #8862?

Reply:
Perhaps. Later I'll go through the whole file to see how many of `IsIntegral` can be removed.

Reply:
Unfortunately, the zero polynomial always splits ([Polynomial.splits_zero](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Data/Polynomial/Splits.html#Polynomial.splits_zero)). I don't know why people special cased 0 in the first place, maybe it's related to that Polynomial.roots 0 = 0 (this has to be special cased as a multiset can't be infinite). If you want, you may try to remove the special case and see what happens.

Reply:
> Unfortunately, the zero polynomial always splits

Oh OK 😅 I made a mistake then.

Reply:
I propose the following code, extracting from #8862:
```lean
theorem Polynomial.Separable.isIntegral {A B : Type*} [CommRing A] [Ring B] [Algebra A B]
    {x : B} (h : (minpoly A x).Separable) : IsIntegral A x := by
  cases subsingleton_or_nontrivial A
  · haveI := Module.subsingleton A B
    exact ⟨1, monic_one, Subsingleton.elim _ _⟩
  · exact of_not_not fun h' ↦ not_separable_zero (minpoly.eq_zero h' ▸ h)
```

Then I can write:
```lean
example {A B : Type*} [CommRing A] [Ring B] [Algebra A B]
    {x : B} (h : (minpoly A x).Separable) : True := by
  have := h.isIntegral
  sorry
```

What do you think?

Reply:
Submitted this to #9134.

Reply:
Looks great and I'm glad it's already taken care of!